Fix issues in Cryptocell 310 ccm_alt discovered by On Target Testing

RonEld opened the pull request:
### Description

<!-- 
    Required
    Add here detailed changes summary, testing results, dependencies 
    Good example: https://os.mbed.com/docs/latest/reference/workflow.html (Pull request template)
-->
Fix issues discovered by On Target Testing on the NRF52840_DK platform:
~1. Check add len is not too big.~
2. Fix memory overflow, by adding a local buffer for the tag result.
3. Return `MBEDTLS_ERR_CCM_AUTH_FAILED` where needed.
4. Add unsupported functions for CCM*.
5. Zeroize output buffer, upon authentication faliure.
6. Use `mbedtls_platform_zeroize()` instead of `mbedtls_zeroize()`


### Pull request type

<!-- 
    Required
    Please add only one X to one of the following types. Do not fill multiple types (split the pull request otherwise).
    Please note this is not a GitHub task list, indenting the boxes or changing the format to add a '.' or '*' in front
    of them would change the meaning incorrectly. The only changes to be made are to add a description text under the
    description heading and to add a 'x' to the correct box.
-->
    [x] Fix
    [ ] Refactor
    [ ] Target update
    [ ] Functionality change
    [ ] Docs update
    [ ] Test update
    [ ] Breaking change

RonEld commented:
Note: This PR will conflict with #8643

0xc0170 commented:
>Note: This PR will conflict with #8643

What order they should be integrated? I assume first 8643 and this one rebased afterwards?

RonEld commented:
@0xc0170 Actually they are not dependent on each other.
The conflict will be because #8643 renames error codes returned by the function, while this PR, returns the previous error code, but added some conditions, in the same changed lines.

This PR fixes *exiting* bugs, while 8643 resolves future issues ( after newer version of Mbed TLS will be merged)
In summary, I would first integrate this PR, and after that rebase 9643
Note that 9643 is not a blocker, as merging Mbed TLS should not break build, as it was first suspected, and was the incentive for that PR

adbridge reviewed (changes requested):
There appear to be camelcase variables in this code. This contravenes the mbed-os coding guidelines....

0xc0170 commented:
> There appear to be camelcase variables in this code. This contravenes the mbed-os coding guidelines....

From my review, those are the one from SDK, not our functions, so should be good. @RonEld Please recheck

0xc0170 commented:
@ARMmbed/mbed-os-crypto Please review the latest update

RonEld commented:
I wil change all the camelcase variables that are changeable:
`CrysRet` -> `crys_ret`
`CC_Mac_Res` -> `cc_mac_res`

Commit:
Style fixes

1. Change camelcase variables to Mbed OS style.
2. Remove functions declarations from the `_alt` header, since they
are now added from the module header regardless whether an alternative
implementation exists.
3. Remove the `extern "c"` declaration from the `_alt` headers.
4. Remove whitespaces before opening parenthesis.
5. Fix alignment of function parameters.
6. Fix indentations.
7. Limit lines to 80 characters.

RonEld commented:
@0xc0170 @adbridge I have renamed the camelcase variable.
I have also modified the header files and remove the function declarations, as they are now not needed to be declared in the alternative headers

yanesca commented:
@RonEld The first commit fixes 6 unrelated issues. Could you please split it up to 6 different commits?

Commit:
Add unsupported functions for CCM*.

Implement the CCM* functions, by returning `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE`

Commit:
Use mbedtls_platform_zeroize

Rplace calls to `mbedtls_zeroize()` to `mbedtls_platform_zeroize()`.

Commit:
Fix memory overflow

Fix memory overflow, by adding a local buffer of type
`CRYS_AESCCM_Mac_Res_t `for the tag result, and copy to the input
parameter `tag`.

Commit:
Additional fix for ccm_alt from On Target Testings

An additional fix for ccm_alt, when the message length is too big.
Check for specific error returned from CC310, and return
`MBEDTLS_ERR_CCM_BAD_INPUT`.

Commit:
Return `MBEDTLS_ERR_CCM_AUTH_FAILED` where needed.

On authentication decryption, when the driver returns the relevant error,
return `MBEDTLS_ERR_CCM_AUTH_FAILED`.

Commit:
Zeroize output buffer, upon authentication faliure.

Clear output buffer upon failure.

0xc0170 commented:
Updated, @yanesca please rereview

RonEld commented:
@yanesca I rebased to split the first commit to 6 different commits.
I also amended the "Additional fix for ccm_alt from On Target Testings" commit with the same fix on the encryption path, which was lost among all the merges \ rebases.
I also amended the style fixes commit with removing whitespaces before the opening parenthesis.

yanesca commented:
Thank you very much, I'll have a look at it as soon as I can!

RonEld commented:
@yanesca Thanks!
I've been thinking. The function `mbedtls_ccm_setkey()`, returns `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE` for every key size that is not 128 bits long, because Cryptocell310 only supports this key size. I think it is worth returning `MBEDTLS_ERR_CCM_BAD_INPUT` in invalid key sizes ( such as 224), and return the `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE` for 192 and 256 bits size keys. What do you think? If you agree, I will work on this first thing next week, but I don't think it should block this PR

yanesca commented:
I agree, that would be a nice improvement!

Commit:
Return MBEDTLS_ERR_CCM_BAD_INPUT on invalid key

Return `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE` only for valid key lengths,
that are not supported by Cryptocell 310. For other key sizes, return
`MBEDTLS_ERR_CCM_BAD_INPUT`

RonEld commented:
@yanesca I have added this improvement. please review

yanesca commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -136,4 +136,22 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+                         const unsigned char *input, unsigned char *output,
+                         unsigned char *tag, size_t tag_len )
+{
+    return( MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE );
```
I know that this wasn't the case when you submitted the PR, but now `MBEDTLS_ERR_AES_FEATURE_UNAVAILABLE` is deprecated and `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` should be used. Would you like to fix it now or would you rather do it together with the others?

RonEld replied:
This PR is not dependent on any preceding PR, and can be merged as soon as it is approved. It fixes *existing* bugs.
The change to the new error, `MBEDTLS_ERR_PLATFORM_FEATURE_UNSUPPORTED` is done in #8643 .
Once this PR will be merged(and #8728), I will rebase #8643 to resolve conflicts, and update this error top the new one.

yanesca commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -84,8 +84,11 @@ int mbedtls_ccm_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
     if( iv_len < 7 || iv_len > 13 )
         return ( MBEDTLS_ERR_CCM_BAD_INPUT );
 
+    if( add_len > 0xFF00 )
```
There is something off: commit message says `tag_len` and the `add_len` is being checked here. What is the intention with this check?

RonEld replied:
It's a typo in the commit message. I'll amend the message

yanesca replied:
Before you do that, could you please explain why this check is necessary?

RonEld replied:
It was raised by the [test case](https://github.com/ARMmbed/mbedtls/blob/development/tests/suites/test_suite_ccm.data#L38)
And check taken from the [Mbed TLS code](https://github.com/ARMmbed/mbedtls/blob/development/library/ccm.c#L165)

yanesca replied:
This is an Mbed TLS specific limitation, we shouldn't constrain hardware accelerators that might have full support. We should run this test only for the Mbed TLS implementation:
https://github.com/ARMmbed/mbedtls/issues/1996

RonEld replied:
Thanks, I'll remove this commit from the PR

yanesca reviewed (commented):
Looks good to me in general, but I have a couple of questions, before I could approve.

RonEld commented:
@yanesca I removed the commit
Please review

Commit:
Add the platform generic errors

Replace the module specific errors with the `MBEDTLS_ERR_PLATFORM_XXX`
errors.

RonEld commented:
Now that #8859 has been merged, I have rebased and changed the module specific errors to the generic `MBEDTLS_ERR_PLATFORM_XXX` errors, to (hopefully) avoid conflicts with #8643 

cc @yanesca

0xc0170 commented:
> Now that #8859 has been merged, I have rebased and changed the module specific errors to the generic MBEDTLS_ERR_PLATFORM_XXX errors, to (hopefully) avoid conflicts with #8643

Conflicts now, please rebase.

Still waiting for @yanesca review?

0xc0170 commented:
targeting 5.11?

yanesca reviewed (approved):
Looks good to me!

@dgreen-arm will have a look at it too, once we have his approval, we are ready to go.

RonEld commented:
@0xc0170 I would like to target 5.11, as it fixes existing issues

RonEld commented:
@yanesca Would you like to wait for the rebase?

dgreen-arm commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -72,30 +80,46 @@ int mbedtls_ccm_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
-    if( CrysRet != CRYS_OK )
-        return ( MBEDTLS_ERR_CCM_HW_ACCEL_FAILED );
+    crys_ret =  CRYS_AESCCM( SASI_AES_ENCRYPT, ctx->cipher_key, ctx->key_size, (uint8_t*)iv, iv_len,
+                            (uint8_t*)add, add_len,  (uint8_t*)input, length, output, tag_len, cc_mac_res );
```
Indentation is a bit off here.

dgreen-arm commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -72,30 +80,46 @@ int mbedtls_ccm_encrypt_and_tag( mbedtls_ccm_context *ctx, size_t length,
+    if( crys_ret == CRYS_AESCCM_ILLEGAL_PARAMETER_SIZE_ERROR )
+    {
+       ret = MBEDTLS_ERR_CCM_BAD_INPUT;
+       goto exit;
```
Indentation is incorrect here.

dgreen-arm commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -109,31 +133,76 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+        * returning this error code, and we don't want to overflow
+        * the output buffer.
+        */
+       return( MBEDTLS_ERR_CCM_BAD_INPUT );
```
Indentation again.

dgreen-arm commented on features/cryptocell/FEATURE_CRYPTOCELL310/ccm_alt.c:
```diff
@@ -109,31 +133,76 @@ int mbedtls_ccm_auth_decrypt( mbedtls_ccm_context *ctx, size_t length,
+                      const unsigned char *iv, size_t iv_len,
+                      const unsigned char *add, size_t add_len,
+                      const unsigned char *input, unsigned char *output,
+                      const unsigned char *tag, size_t tag_len )
```
The parameter alignment seems wrong here in these two functions

RonEld replied:
ok, but this is same is in the [origin](https://github.com/ARMmbed/mbedtls/blob/development/library/ccm.c#L307)

dgreen-arm replied:
Probably okay to leave it then.

RonEld replied:
no problem, I'll fix it

dgreen-arm reviewed (commented):
Some minor indentation and alignment issues, but looks good on the whole.

0xc0170 commented:
@RonEld Please also rebase with the fixing the review above

Once done, we'll start CI

RonEld commented:
@0xc0170 Sure, no problem.
Since the fixes are style changes, would you like me to squash the fixes with https://github.com/ARMmbed/mbed-os/pull/8704/commits/c314242a7717298676cb9e80738ff76f24eb2b85 ?

0xc0170 commented:
> Since the fixes are style changes, would you like me to squash the fixes with c314242 ?

It could be.

RonEld commented:
@0xc0170 I have rebased to resolve conflicts, and put fixes from review from @dgreen-arm in https://github.com/ARMmbed/mbed-os/pull/8704/commits/1b34927e83fc502ee01d139ed0c0947c870d4700

yanesca commented:
@RonEld do you still have the branch before the rebase? Can you post a link to it so that I can compare?

bulislaw commented:
All the PRs need to be engineering ready (marked as "needs: CI") by the end of the day (Austin time). Otherwise it won't make 5.11 and will need to come in the next release (5.12 for features, 5.11.1 for fixes and new platforms).

RonEld commented:
@yanesca UnfortunatelyI don't.. I am working on making a reference branch

RonEld commented:
@yanesca I believe the branch before the rebase and style fixes from @dgreen-arm is in https://github.com/RonEld/mbed-os/tree/ccm_alt_reference_brach

yanesca reviewed (approved):
I have reviewed the rebase and I am happy with the changes.

(@RonEld thanks for putting up a reference branch!)

cmonr commented:
@ARMmbed/mbed-os-maintainers _IF_ we're ignoring the `jenkins_ci/exporter` result (the export nodes are currently completely out of space and failing), then this PR is good to come in.

0xc0170 commented:
Thanks @cmonr for heads-up! This is ready for merge
